### PR TITLE
Implement PluginSearch TextBox & ContentControl Lazy Load

### PR DESCRIPTION
**Pull request by taooceros**
- Plugin store search
- Ctrl + F when in the plugin store page to focus cursor on the search box
- Plugins tab search

### CODE REVIEW

**Comment by jjw24** [comment]
looks like it's still WIP? i will mark it as draft

File: Flow.Launcher/ViewModel/SettingWindowViewModel.cs [hunk 1]
```diff
@@ -19,6 +19,7 @@
 using Flow.Launcher.Infrastructure.UserSettings;
 using Flow.Launcher.Plugin;
 using Flow.Launcher.Plugin.SharedModels;
+using System.Windows.Data;
```
Comment (jjw24) [inline Flow.Launcher/ViewModel/SettingWindowViewModel.cs:22]: Not needed
```suggestion
```

**Comment by jjw24** [comment]
search is not working, tried pressing enter too.

**Comment by jjw24** [comment]
Tried it again and still the Plugin Store's search is not working. I removed entire flow user data then rebuilt and ran. Went into plugin store tab and clicked on the search text, put in a search term 'colors' and pressed enter. Nothing happens.

Search is however working for me in the plugins tab.

Also:
- can we add Ctrl + F focus to the Plugin Store search text box please
- add a tooltip to both tabs so when hovering over the search text box it displays something like 'Ctrl + F to search plugins'

**Comment by taooceros** [comment]
> Tried it again and still the Plugin Store's search is not working. I removed entire flow user data then rebuilt and ran. Went into plugin store tab and clicked on the search text, put in a search term 'colors' and pressed enter. Nothing happens.
> 
> Search is however working for me in the plugins tab.
> 
> Also:
> 
>     * can we add Ctrl + F focus to the Plugin Store search text box please
> 
>     * add a tooltip to both tabs so when hovering over the search text box it displays something like 'Ctrl + F to search plugins'

Ah I get your idea🤣I haven't implemented the plugin store search. I will fix that later.

**Comment by onesounds** [comment]
- Plugin Store works. Plugin list doesn't works. 
- It is good to be searched when entering the enter key, but if the search box is empty, it is necessary to initialize it.

- Can we add label? (installed, update, etc...) if so, we need more works with install button. (update, disabled...)
- other topic. we need something to message when no result / no connection.

**Comment by taooceros** [comment]
> It is good to be searched when entering the enter key, but if the search box is empty, it is necessary to initialize it.

That's pretty slow though, because empty means there are most plugins that needs to be rendered. Though it makes sense. We can add a LostFocus event for that.

**Comment by jjw24** [comment]
this is ready for review?

**Comment by onesounds** [comment]
> this is ready for review?

yep. There are 'installed' labels or a few more features I want to add, but I think it would be better to do it in another pr. The search itself works well.

File: Flow.Launcher.sln [hunk 2]
```diff
@@ -213,7 +213,6 @@ Global
 		{C21BFF9C-2C99-4B5F-B7C9-A5E6DDDB37B0}.Release|x86.ActiveCfg = Release|Any CPU
 		{C21BFF9C-2C99-4B5F-B7C9-A5E6DDDB37B0}.Release|x86.Build.0 = Release|Any CPU
 		{9B130CC5-14FB-41FF-B310-0A95B6894C37}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
-		{9B130CC5-14FB-41FF-B310-0A95B6894C37}.Debug|Any CPU.Build.0 = Debug|Any CPU
```
Comment (jjw24) [inline Flow.Launcher.sln:216]: why is this removed?

File: Flow.Launcher/SettingWindow.xaml [hunk 3]
```diff
@@ -323,8 +323,8 @@
             <Setter Property="Padding" Value="0,0,0,0" />
             <Setter Property="Margin" Value="0,0,8,5" />
             <Setter Property="BorderBrush" Value="{DynamicResource Color03B}" />
-            <Setter Property="MinWidth" Value="330" />
-            <Setter Property="Height" Value="98" />
+            <Setter Property="MinWidth" Value="100" />
+            <Setter Property="Height" Value="Auto" />
```
Comment (jjw24) [inline Flow.Launcher/SettingWindow.xaml:327]: Do we need to change this? ATM when you open settings for the first time on a brand new install, the Settings window is opening to the top left corner instead of center.

File: Flow.Launcher/Languages/en.xaml [hunk 4]
```diff
@@ -60,6 +60,11 @@
     <system:String x:Key="shadowEffectNotAllowed">Shadow effect is not allowed while current theme has blur effect enabled</system:String>
 
     <!--  Setting Plugin  -->
+    <system:String x:Key="searchplugin">Search Plugin</system:String>
+    <system:String x:Key="searchpluginToolTip">Ctrl+F to Search Plugins</system:String>
+    <system:String x:Key="searchplugin_Noresult_Title">No results to display</system:String>
+    <system:String x:Key="searchplugin_Noresult_Subtitle">Try another search term.</system:String>
+    <system:String x:Key="searchplugin_Noresult_Subtitle_Network">Please check your network connection or try another search term.</system:String>
```
Comment (jjw24) [inline Flow.Launcher/Languages/en.xaml:67]:
```suggestion
    <system:String x:Key="searchplugin_Noresult_Subtitle_Network">No results found. Please try a different search.</system:String>
```

Comment (jjw24) [inline Flow.Launcher/Languages/en.xaml:67]: These different states, I can't seem to hit the 'searchplugin_Noresult_Subtitle', and displaying about network connection is confusing no?
image

Comment (jjw24) [reply Flow.Launcher/Languages/en.xaml:67]: Should it be Plugin Store on load if no results -> 'No plugins can be loaded, please check your network connection.'
When searching in Plugins tab or Plugin Store and no results found -> 'No results found. Please try a different search.'

File: Flow.Launcher/SettingWindow.xaml.cs [hunk 5]
```diff
@@ -361,13 +370,83 @@ private void Window_StateChanged(object sender, EventArgs e)
...
+        {
+            if (string.IsNullOrEmpty(pluginFilterTxb.Text))
+                return true;
+            if (item is PluginViewModel model)
+            {
+                return StringMatcher.FuzzySearch(pluginFilterTxb.Text, model.PluginPair.Metadata.Name).IsSearchPrecisionScoreMet();
+            }
+            return false;
+        }
+        private bool PluginStoreFilter(object item)
```
Comment (jjw24) [inline Flow.Launcher/SettingWindow.xaml.cs:386]:
```suggestion

        private bool PluginStoreFilter(object item)
```

File: Flow.Launcher/SettingWindow.xaml.cs [hunk 6]
```diff
@@ -361,13 +370,84 @@ private void Window_StateChanged(object sender, EventArgs e)
...
+            if (e.Key == Key.Enter)
+                RefreshPluginListEventHandler(sender, e);
+        }
+
+        private void PluginStoreFilterTxb_OnKeyDown(object sender, KeyEventArgs e)
+        {
+            if (e.Key == Key.Enter)
+                RefreshPluginStoreEventHandler(sender, e);
+        }
+        private void OnPluginSettingKeydown(object sender, KeyEventArgs e)
```
Comment (jjw24) [inline Flow.Launcher/SettingWindow.xaml.cs:430]:
```suggestion

        private void OnPluginSettingKeydown(object sender, KeyEventArgs e)
```

File: Flow.Launcher/SettingWindow.xaml.cs [hunk 7]
```diff
@@ -361,13 +370,85 @@ private void Window_StateChanged(object sender, EventArgs e)
...
         private void ItemSizeChanged(object sender, SizeChangedEventArgs e)
         {
-            Plugins.ScrollIntoView(Plugins.SelectedItem);
+
+        }
+
+        private void SelectedPluginChanged(object sender, SelectionChangedEventArgs e)
+        {
+
+        }
```
Comment (jjw24) [inline Flow.Launcher/SettingWindow.xaml.cs:445]: Are these empty methods needed?

File: Flow.Launcher/ViewModel/PluginViewModel.cs [hunk 8]
```diff
@@ -30,9 +30,20 @@ public bool PluginState
...
+                _isExpanded = value;
+                OnPropertyChanged();
+                OnPropertyChanged(nameof(SettingControl));
+            }
+        }
 
         private Control _settingControl;
-        public Control SettingControl => _settingControl ??= PluginPair.Plugin is not ISettingProvider settingProvider ? new Control() : settingProvider.CreateSettingPanel();
+        private bool _isExpanded;
+        public Control SettingControl => IsExpanded ? _settingControl ??= PluginPair.Plugin is not ISettingProvider settingProvider ? new Control() : settingProvider.CreateSettingPanel() : null;
```
Comment (jjw24) [inline Flow.Launcher/ViewModel/PluginViewModel.cs:46]: @taooceros this statement is doing my head in, what's it for?

File: Flow.Launcher/SettingWindow.xaml (hunk 3 above)
Comment (onesounds) [reply Flow.Launcher/SettingWindow.xaml]: Useless Value. Removed.

File: Flow.Launcher.sln (hunk 2 above)
Comment (onesounds) [reply Flow.Launcher.sln]: I guess I turned off. Recovered.

File: Flow.Launcher/Languages/en.xaml (hunk 4 above)
Comment (onesounds) [reply Flow.Launcher/Languages/en.xaml]: I simply removed network part, and changed text to 'No results found. Please try a different search.'

Since the network connection verification function is not implemented (I remember that Chinese users have left an issue here) I wanted to put in a simple message. Even when it was not connected, the result was zero, so I tried to put a warning message here. I thought it was a good idea because I didn't need a function code. Let's implement and add the network confirmation part in a separate PR.

File: Flow.Launcher/ViewModel/PluginViewModel.cs (hunk 8 above)
Comment (taooceros) [reply Flow.Launcher/ViewModel/PluginViewModel.cs:46]: Unless the panel is expanded, let's not load the usercontrol.

File: Flow.Launcher/SettingWindow.xaml.cs (hunk 7 above)
Comment (taooceros) [reply Flow.Launcher/SettingWindow.xaml.cs:445]: I forget, but if it is empty, I think not needed.

**Review by jjw24 — APPROVED** [review]
